### PR TITLE
Lazy load `torch` in surrogates

### PR DESCRIPTION
**Pull request by rjavadi**
`baybe.utils.numerical` imports torch and still appears in the import hierarchy. I'm going to fix it in another PR.

Screenshot 2024-03-06 at 4 44 39 PM

### CODE REVIEW

File: baybe/recommenders/pure/bayesian/base.py
Comment (rjavadi) [inline baybe/recommenders/pure/bayesian/base.py:1]: This change is not strictly related to surrogates. Please let me know If you think it shouldn't be in this PR.

File: baybe/serialization/mixin.py
Comment (rjavadi) [inline baybe/serialization/mixin.py:1]: As we discussed: https://github.com/emdgroup/baybe/discussions/162

File: baybe/utils/basic.py
Comment (rjavadi) [inline baybe/utils/basic.py:1]: Since it appeared in the import hierarchy of `import surrogates` I loaded torch lazily here as well.

File: baybe/recommenders/pure/bayesian/base.py
Comment (AVHopp) [reply baybe/recommenders/pure/bayesian/base.py:1]: You mean the change of where the acquisition functions are loaded? This comment is shown at the top of the file for me, so I cannot identify whether your comment just relates to a special part of the changes here

File: baybe/surrogates/utils.py
[the remark this answers is not in the document]
Comment (AVHopp) [reply baybe/surrogates/utils.py:1]: What about @Scienfitz proposal of making a dedicated file `baybe.utils.torch`? That was also part of that discussion but I think this point was not discussed yet.

**Review by AVHopp — COMMENTED** [review]
Depending on the decision regarding a `baybe.utils.torch` file, I'll either hit "Approve" or re-evaluate. But I think this already looks good :)

File: baybe/recommenders/pure/bayesian/base.py
Comment (rjavadi) [reply baybe/recommenders/pure/bayesian/base.py:1]: Yes, according to this screenshot Adrian sent in our meeting:
Bild

File: baybe/surrogates/utils.py
[the remark this answers is not in the document]
Comment (rjavadi) [reply baybe/surrogates/utils.py:1]: Yes, I can do that in this PR.

[the remark this answers is not in the document]
Comment (rjavadi) [reply baybe/surrogates/utils.py:1]: Adding `baybe.utils.torch` now might introduce unnecessary complexity to this pull request. Let's merge first and then we can design `baybe.utils.torch` to better integrate with the codebase.

[the remark this answers is not in the document]
Comment (rjavadi) [reply baybe/surrogates/utils.py:1]: @AVHopp @Scienfitz Also, it would be very helpful if you share your "wishlist" about `baybe.utils.torch`. What do you think should be included there other than `numerical.DTypeFloatTorch`?

File: baybe/recommenders/pure/bayesian/base.py
Comment (AVHopp) [reply baybe/recommenders/pure/bayesian/base.py:1]: Ok, just wanted to make sure :)

File: baybe/surrogates/utils.py
[the remark this answers is not in the document]
Comment (AVHopp) [reply baybe/surrogates/utils.py:1]: I personally do not have a strong opinion. I do not think that we desperately need such a file, but depending on the planned content of that file it might make sense.
I'll leave discussion and decision this to the two of you.

**Review by AVHopp — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -12,7 +17,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - `from_simplex` constructor now also takes and applies optional constraints
 
 ### Changed
-- Full lookup backtesting example now tests different substance encodings
+- Full lookup backtesting example now tests different substance encodings
```
Comment (Scienfitz) [inline CHANGELOG.md:20]: space added intentionally?

File: baybe/recommenders/pure/bayesian/base.py [hunk 2]
```diff
@@ -53,17 +42,8 @@ def _get_acquisition_function_cls(
...
-            "PI": ProbabilityOfImprovement,
-            "EI": ExpectedImprovement,
-            "UCB": partial(UpperConfidenceBound, beta=1.0),
-            "qEI": qExpectedImprovement,
-            "qPI": qProbabilityOfImprovement,
-            "qUCB": partial(qUpperConfidenceBound, beta=1.0),
-            "VarUCB": partial(UpperConfidenceBound, beta=100.0),
-            "qVarUCB": partial(qUpperConfidenceBound, beta=100.0),
-        }
+        from baybe.acquisition import debotorchize, mapping
```
Comment (Scienfitz) [inline baybe/recommenders/pure/bayesian/base.py:45]: trying to undertsand why youre doing thes echange sin this file, so feel free to elaborate:

at the top of the file you still import something from botorch, hence torch. So this lazy import would have 0 impact, or am I missing something? Or should the `AcquisitionFunction` import also be lazy/type-checking only?

From another angle, this file is part of the bayesian recommender submodule. Anything bayesian (since we use botorch for it) will always have to do with torch/botorch via the acquisition functions, hence why the move here to import it lazily?

File: baybe/serialization/mixin.py [hunk 3]
```diff
@@ -3,6 +3,8 @@
 import json
 from typing import Type, TypeVar
 
+import attr
```
Comment (Scienfitz) [inline baybe/serialization/mixin.py:6]: im not 100% up to date why theres still something called `attr` but for consistency we always should use the `attrs` version of stuff (unless theres a meaningful difference, which doesnt seem the case here)

File: baybe/serialization/mixin.py [hunk 4]
```diff
@@ -29,7 +31,11 @@ def from_dict(cls: Type[_T], dictionary: dict) -> _T:
         Returns:
             The reconstructed object.
         """
-        return converter.structure(dictionary, cls)
+        # In case we use postponed annotations, we must resolve the original type
```
Comment (Scienfitz) [inline baybe/serialization/mixin.py:34]: the note says `in case` but theres no case distinction here (a la if else or similar), does this also cover the previous version and merely extends it?

File: baybe/surrogates/custom.py [hunk 5]
```diff
@@ -7,12 +7,11 @@
 
 It is planned to solve this issue in the future.
 """
+from __future__ import annotations
```
Comment (Scienfitz) [inline baybe/surrogates/custom.py:10]: I dont immediately understand why this has become needed in several places where you've added it, can you elaborate?

File: baybe/surrogates/utils.py
[the remark this answers is not in the document]
Comment (Scienfitz) [reply baybe/surrogates/utils.py:1]: the wishlist would prob consist of all utilities and stuff like the dtype that require torch, then wed have one single utility file that needs torch and none of the others would need it

might be that theres mixed utilities that would not make sense moving (like eg the ranodm seed setter) so Im also open to not having such a file if it would have no content or breaking of other logic groupings

File: baybe/acquisition.py [hunk 6]
```diff
@@ -1,18 +1,43 @@
...
 from botorch.posteriors import Posterior
 from botorch.posteriors.gpytorch import GPyTorchPosterior
-from torch import Tensor, cat, squeeze
 
 from baybe.surrogates.base import Surrogate
 
+if TYPE_CHECKING:
+    from torch import Tensor
+
+mapping = {
```
Comment (Scienfitz) [inline baybe/acquisition.py:29]: since this is now not a function-local variable anymore Id give it a more precise name, ie mapping of what?

File: baybe/acquisition.py [hunk 7]
```diff
@@ -1,18 +1,43 @@
...
+)
 from botorch.models.gpytorch import Model
 from botorch.posteriors import Posterior
 from botorch.posteriors.gpytorch import GPyTorchPosterior
-from torch import Tensor, cat, squeeze
 
 from baybe.surrogates.base import Surrogate
 
+if TYPE_CHECKING:
+    from torch import Tensor
```
Comment (Scienfitz) [inline baybe/acquisition.py:27]: again the question about whether this even makes sense here? a lot of plain botorch imorts above... wouldnt that mean the lazy torch import is implactless??

File: CHANGELOG.md (hunk 1 above)
Comment (rjavadi) [reply CHANGELOG.md:20]: No

File: baybe/serialization/mixin.py (hunk 4 above)
Comment (rjavadi) [reply baybe/serialization/mixin.py:34]: Yes, that's an extension to handle postponed annotations.

File: baybe/surrogates/custom.py (hunk 5 above)
Comment (rjavadi) [reply baybe/surrogates/custom.py:10]: Because we don't want to use string types for `Tensor` which is imported only for type checking and not runtime.
For example, we should write `def _posterior(self, candidates: Tensor) -> Tuple[Tensor, Tensor]` like:
`def _posterior(self, candidates: "Tensor") -> Tuple["Tensor", "Tensor"]`

File: baybe/recommenders/pure/bayesian/base.py (hunk 2 above)
Comment (rjavadi) [reply baybe/recommenders/pure/bayesian/base.py:45]: Here's the hierarchy of imports for `base.py`:

Screenshot 2024-03-13 at 6 18 27 PM

torch only appears in the hierarchy of `baybe.utils.numericals`.

File: baybe/acquisition.py (hunk 7 above)
Comment (rjavadi) [reply baybe/acquisition.py:27]: It's similar to `baybe/recommenders/pure/bayesian/base.py`:

Screenshot 2024-03-13 at 6 29 11 PM

File: baybe/surrogates/custom.py (hunk 5 above)
Comment (Scienfitz) [reply baybe/surrogates/custom.py:10]: and why was that not a problem before hence requiring the change?

File: baybe/acquisition.py (hunk 7 above)
Comment (Scienfitz) [reply baybe/acquisition.py:27]: i cant see botorch in that picture despite its import at top of file, can you check?
also, the picture just shows the trace hence the first import of torch, not all traces that would trigger torch import

File: baybe/recommenders/pure/bayesian/base.py (hunk 2 above)
Comment (rjavadi) [reply baybe/recommenders/pure/bayesian/base.py:45]: Yes, you're right regarding the torch bubble. I'm gonna discard unrequired changes.

File: baybe/acquisition.py (hunk 7 above)
Comment (rjavadi) [reply baybe/acquisition.py:27]: Yes, that's right. I truncated image for readability. `botorch` appears here:
Screenshot 2024-03-15 at 7 06 24 AM

File: baybe/acquisition.py [hunk 8]
```diff
@@ -1,18 +1,39 @@
...
+)
 from botorch.models.gpytorch import Model
 from botorch.posteriors import Posterior
 from botorch.posteriors.gpytorch import GPyTorchPosterior
-from torch import Tensor, cat, squeeze
+from torch import Tensor
 
 from baybe.surrogates.base import Surrogate
 
+acquisition_function_mapping = {
```
Comment (rjavadi) [inline baybe/acquisition.py:25]: @Scienfitz Here I renamed `mapping` to `acquisition_function_mapping`.

File: baybe/acquisition.py [hunk 9]
```diff
@@ -1,18 +1,39 @@
...
+    UpperConfidenceBound,
+    qExpectedImprovement,
+    qProbabilityOfImprovement,
+    qUpperConfidenceBound,
+)
 from botorch.models.gpytorch import Model
 from botorch.posteriors import Posterior
 from botorch.posteriors.gpytorch import GPyTorchPosterior
-from torch import Tensor, cat, squeeze
+from torch import Tensor
```
Comment (rjavadi) [inline baybe/acquisition.py:21]: Added `torch` to imports because here we're in torch-bubble.

File: baybe/acquisition.py (hunk 8 above)
Comment (Scienfitz) [reply baybe/acquisition.py:26]: ty

File: baybe/acquisition.py (hunk 6 above)
Comment (Scienfitz) [reply baybe/acquisition.py]: see other comment

**Review by Scienfitz — APPROVED** [review]

**Comment by AdrianSosic** [comment]
Delay `torch` import until really needed
